### PR TITLE
Allow multiple publishing on multiple interfaces on same port.

### PR DESCRIPTION
**Pull request by TortleWortle**
Fixes the issue where you cannot have multiple ip addresses on the same port for a server.

### CODE REVIEW

File: environment/allocations.go
Comment (matthewpi) [inline environment/allocations.go]: `if tcpPort, ok := out[nat.Port(fmt.Sprintf("%d/tcp", port))]; !ok {`

File: environment/allocations.go
Comment (matthewpi) [inline environment/allocations.go]: `if udpPort, ok := out[nat.Port(fmt.Sprintf("%d/udp", port))]; !ok {`

**Review by matthewpi — CHANGES_REQUESTED** [review]

File: environment/allocations.go
Comment (TortleWortle) [reply environment/allocations.go]: good catch, thanks

File: environment/allocations.go
Comment (DaneEveritt) [inline environment/allocations.go]: `if v, ok := ...; ok {` would be more in-line stylistically (and logically) with the rest of the code base. I don't like doing the negated check as the primary statement here, it was confusing because in my head the negated should be in the else block.

It reads more like "if there is a value, append it, otherwise create the binding element" rather than what you have which is "if there is not a value create it, otherwise append it." Perhaps that is just my brain though.

**Review by DaneEveritt — CHANGES_REQUESTED** [review]

File: environment/allocations.go
Comment (TortleWortle) [reply environment/allocations.go]: Agreed, I initially did it that way but changed it because @matthewpi's comment had `!ok`.

How about this:
```go
binding := nat.PortBinding{
	HostIP:   ip,
	HostPort: strconv.Itoa(port),
}

tcpKey := nat.Port(fmt.Sprintf("%d/tcp", port))
udpKey := nat.Port(fmt.Sprintf("%d/udp", port))

if tcpPort, ok := out[tcpKey]; ok {
	out[tcpKey] = append(tcpPort, binding)
} else {
	out[tcpKey] = []nat.PortBinding{binding}
}

if udpPort, ok := out[udpKey]; ok {
	out[udpKey] = append(udpPort, binding)
} else {
	out[udpKey] = []nat.PortBinding{binding}
}
```

Comment (matthewpi) [reply environment/allocations.go]: Yeah flip the `ok`, I just left it how it was because I was ignoring that there was an else case.

Comment (DaneEveritt) [reply environment/allocations.go]: I prefer the `if v, ok :=` (emphasis on the `v`) since it is more idiomatic and since tcpPort/udpPort are just superfluous since they're scoped only to the if/else block as is.

Comment (schrej) [reply environment/allocations.go]: I'd actually go for
```go
if _, ok := out[tcpKey]; ok {
  out[tcpKey] = append(out[tcpKey], binding)
} // ...
```
here, since that's much easier to read imho.

EDIT: thinking about it a little more, it probably does not really matter that much.

File: environment/allocations.go
Comment (schrej) [inline environment/allocations.go]: Those `fmt.Sprintf`s should definitely be extracted into variables as you suggested yourself in one of the review comments.

**Review by schrej — CHANGES_REQUESTED** [review]

File: environment/allocations.go
Comment (matthewpi) [reply environment/allocations.go]: The entire `nat.Port()` should be extracted into variables, not just the `fmt.Sprintf`s

Comment (schrej) [reply environment/allocations.go]: Yes. It's late here.

Comment (TortleWortle) [reply environment/allocations.go]: So together with your previous reply we could make it:
```go
tcp := nat.Port(fmt.Sprintf("%d/tcp", port))
udp := nat.Port(fmt.Sprintf("%d/udp", port))

out[tcp] = append(out[tcp], binding)
out[udp] = append(out[udp], binding)
```

Comment (schrej) [reply environment/allocations.go]: Disregard my comment regarding readability though, I'm not sure what best practice here is, but I guess using the returned value is fine.
So keep it as `if v, ok :=` and `out[tcp] = append(v, binding)`.
You could see if `o` instead of `v` is readable, but it might be bad because of the `ok` afterwards 😄

**Review by matthewpi — APPROVED** [review]
LGTM, once merged I can tweak it if need be.

**Review by DaneEveritt — APPROVED** [review]

**Pull request merged**
